### PR TITLE
[Merged by Bors] - feat: the restriction functor on homological complexes

### PR DESCRIPTION
**Pull request by joelriou**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Homology/Embedding/Restriction.lean [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  (K L M : HomologicalComplex C c') (φ : K ⟶ L) (φ' : L ⟶ M)
+  (e : c.Embedding c') [e.IsRelIff]
+
+/-- Given `K : HomologicalComplex C c'` and `e : c.Embedding c'` (satisfying `[e.IsRelIff]`),
+this is the homological complex in `HomologicalComplex C c` obtained by restriction. -/
+@[simps]
+def restriction : HomologicalComplex C c where
+  X i := K.X (e.f i)
+  d _ _ := K.d _ _
+  shape i j hij := K.shape _ _ (by simpa only [← e.rel_iff] using hij)
```
Comment (callesonne) [inline Mathlib/Algebra/Homology/Embedding/Restriction.lean:34]:
```suggestion
  shape i j hij := K.shape _ _ ((e.rel_iff _ _).1.mt hij)
```

File: Mathlib/Algebra/Homology/Embedding/Restriction.lean [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+/-- The isomorphism `(K.restriction e).X i ≅ K.X i'` when `e.f i = i'`. -/
+def restrictionXIso {i : ι} {i' : ι'} (h : e.f i = i') :
+    (K.restriction e).X i ≅ K.X i' :=
+  eqToIso (by subst h; rfl)
```
Comment (callesonne) [inline Mathlib/Algebra/Homology/Embedding/Restriction.lean:39]:
```suggestion
  eqToIso (h ▸ rfl)
```

File: Mathlib/Algebra/Homology/Embedding/Restriction.lean [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+lemma restrictionMap_f' {i : ι} {i' : ι'} (hi : e.f i = i') :
+    (restrictionMap φ e).f i = (K.restrictionXIso e hi).hom ≫
+      φ.f i' ≫ (L.restrictionXIso e hi).inv := by
+  subst hi
+  simp [restrictionXIso]
+
+variable (K)
+
+@[simp]
+lemma restrictionMap_id : restrictionMap (𝟙 K) e = 𝟙 _ := by aesop_cat
```
Comment (callesonne) [inline Mathlib/Algebra/Homology/Embedding/Restriction.lean:65]:
```suggestion
lemma restrictionMap_id : restrictionMap (𝟙 K) e = 𝟙 _ := rfl
```

File: Mathlib/Algebra/Homology/Embedding/Restriction.lean [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+@[simp, reassoc]
+lemma restrictionMap_comp :
+    restrictionMap (φ ≫ φ') e = restrictionMap φ e ≫ restrictionMap φ' e := by aesop_cat
```
Comment (callesonne) [inline Mathlib/Algebra/Homology/Embedding/Restriction.lean:69]:
```suggestion
    restrictionMap (φ ≫ φ') e = restrictionMap φ e ≫ restrictionMap φ' e := rfl
```

File: Mathlib/Algebra/Homology/Embedding/Restriction.lean [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Homology.Embedding.Basic
+import Mathlib.Algebra.Homology.Additive
+
+/-!
+# The restriction functor of an embedding of complex shapes
+
+Given `c` and `c'` complex shapes on two types, and `e : c.Embedding c'`,
+we define the restriction functor
```
Comment (callesonne) [inline Mathlib/Algebra/Homology/Embedding/Restriction.lean:13]:
```suggestion
Given `c` and `c'` complex shapes on two types, and `e : c.Embedding c'` 
(satisfying `[e.IsRelIff]`), we define the restriction functor
```

File: Mathlib/Algebra/Homology/Embedding/Restriction.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/Algebra/Homology/Embedding/Restriction.lean:34]: Thanks for the other suggestions! I would rather keep using `simpa only` which uses the same number of lines (even though slightly more characters!), as it is slightly more readable IMHO.

**Review by callesonne — APPROVED** [review]
Looks good to me!

**Comment by erdOne** [comment]
Thanks! (and thanks Calle!)
maintainer merge

**Comment by kbuzzard** [comment]
Thanks to everyone involved!

bors merge

**Pull request closed without merging**